sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java

```diff
@@ -211,56 +212,70 @@ Mono<StoreResponse> writePrivateAsync(
+                                        false,
+                                        false,
+                                        primaryUri,
+                                        replicaStatusList);
+                                    String value = ex != null ?
+                                        ex
+                                            .getResponseHeaders()
+                                            .get(HttpConstants
+                                                .HttpHeaders
+                                                .WRITE_REQUEST_TRIGGER_ADDRESS_REFRESH) :
+                                        null;
+                                    if (!Strings.isNullOrWhiteSpace(value)) {
+                                        Integer result = Integers.tryParse(value);
+                                        if (result != null && result == 1) {
+                                            startBackgroundAddressRefresh(request);
+                                        }
+                                    }
+                                } catch (Throwable throwable) {
+                                    if (throwable instanceof Error) {
+                                        logger.error("Unexpected failure in handling orig [{}]", t.getMessage(), t);
+                                        logger.error("Unexpected failure in handling orig [{}] : new [{}]", t.getMessage(), throwable.getMessage(), throwable);
+                                        throw (Error) throwable;
+                                    } else {
+                                        // this happens before any retry policy - like for example GoneAndRetryRetryPolicy
+                                        // kicks in - no need to spam warn/error level logs yet
+                                        logger.info("Unexpected failure in handling orig [{}]", t.getMessage(), t);
+                                        logger.info("Unexpected failure in handling orig [{}] : new [{}]", t.getMessage(), throwable.getMessage(), throwable);
+                                    }
+                                }
+                            }
+                        );
+                } catch (Exception unwrappedException) {
+
+                    CosmosException cosmosException = Utils.as(unwrappedException, CosmosException.class);
 
+                    if (cosmosException != null) {
+                        return Mono.error(cosmosException);
+                    }
+
+                    String errorMessage = "Unexpected exception " + unwrappedException.getMessage() + " received while reading from store.";
+                    return Mono.error(new InternalServerErrorException(
+                        com.azure.cosmos.implementation.Exceptions.getInternalServerErrorMessage(errorMessage),
+                        unwrappedException,
+                        HttpConstants.SubStatusCodes.INVALID_RESULT));
```

Comment:
should we use a different substrates code?

Reply:
Good idea!

Reply:
Fixed in next iteration.